activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,160 @@
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k1 << "stay-puft"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = Key.new(["foo", ["bar"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(Key)
+
+        self.new(key)
+      end
+
+      def initialize(key = nil)
+        @cache_key = +""
+        @cache_version = +""
+        @empty_key = true
+        self << key
+      end
+
+      def update(key)
+        if key.respond_to?(:cache_key)
+          store_key(key.cache_key.to_s)
+          store_version(key)
+          return
+        end
+
+        case key
+        when Array
+          key.each do |k|
+            self << k
+          end
+        when Hash
+          key = key.to_a
+          store_version(key)
+          key.sort_by! { |k, _| k.to_s }
+          key.collect! { |k, v| "#{k}=#{v}" }
+
+          key.each do |k|
+            store_key(k)
+          end
+        else
+          store_version(key)
+          store_key(key)
+        end
+      end
+      alias :<< :update
+
+      private
+        def store_key(key)
+          @cache_key << "/" unless @empty_key
+          @empty_version =
```

Comment:
This line seems weird, since there's nothing on the other side?

Reply:
Thanks, fixing